flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,340 @@
+
+  class Node(center: Vector, width: Vector, var children: Seq[Node]) {
+
+    val nodeElements = new ListBuffer[Vector]
+
+    /** for testing purposes only; used in QuadTreeSuite.scala
+      *
+      * @return center and width of the box
+      */
+    def getCenterWidth(): (Vector, Vector) = {
+      (center, width)
+    }
+
+    def contains(queryPoint: Vector): Boolean = {
+      overlap(queryPoint, 0.0)
+    }
+
+    /** Tests if queryPoint is within a radius of the node
+      *
+      * @param queryPoint
+      * @param radius
+      * @return
+      */
+    def overlap(queryPoint: Vector, radius: Double): Boolean = {
+      var count = 0
+      for (i <- 0 to queryPoint.size - 1) {
+        if (queryPoint(i) - radius < center(i) + width(i) / 2 &&
+          queryPoint(i) + radius > center(i) - width(i) / 2) {
+          count += 1
+        }
+      }
+
+      if (count == queryPoint.size) {
+        true
+      } else {
+        false
+      }
+    }
+
+    /** Tests if queryPoint is near a node
+      *
+      * @param queryPoint
+      * @param radius
+      * @return
+      */
+    def isNear(queryPoint: Vector, radius: Double): Boolean = {
+      if (minDist(queryPoint) < radius) {
+        true
+      } else {
+        false
+      }
+    }
+
+    /**
+     * used in error handling when computing minDist to make sure
+     * distMetric is Euclidean or SquaredEuclidean
+     * @param message
+     */
+    case class metricException(message: String) extends Exception(message)
+
+    /**
+     * minDist is defined so that every point in the box
+     * has distance to queryPoint greater than minDist
+     * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+     *
+     * @param queryPoint
+     * @return
+     */
+
+    def minDist(queryPoint: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to queryPoint.size - 1) {
+        if (queryPoint(i) < center(i) - width(i) / 2) {
+          minDist += math.pow(queryPoint(i) - center(i) + width(i) / 2, 2)
+        } else if (queryPoint(i) > center(i) + width(i) / 2) {
+          minDist += math.pow(queryPoint(i) - center(i) - width(i) / 2, 2)
+        }
+      }
+
+      if (distMetric.isInstanceOf[SquaredEuclideanDistanceMetric]) {
+        minDist
+      } else if (distMetric.isInstanceOf[EuclideanDistanceMetric]) {
+        math.sqrt(minDist)
+      } else{
+        throw metricException(s" Error: metric must be Euclidean or SquaredEuclidean!")
+      }
+    }
+
+    /**
+     * Finds which child queryPoint lies in.  node.children is a Seq[Node], and
+     * whichChild finds the appropriate index of that Seq.
+     * @param queryPoint
+     * @return
+     */
+    def whichChild(queryPoint: Vector): Int = {
+      var count = 0
+      for (i <- 0 to queryPoint.size - 1) {
+        if (queryPoint(i) > center(i)) {
+          count += Math.pow(2, queryPoint.size -1 - i).toInt
+        }
+      }
+      count
+    }
```

Comment:
To avoid using `var`, we can rewrite this method.

```scala
def whichChild(queryPoint: Vector): Int = {
  (0 until queryPoint.size).map { i =>
    if (queryPoint(i) > center(i)) {
      Math.pow(2, queryPoint.size - 1 - i).toInt
    } else {
      0
    }
  }.sum
}
```